notus/scanner/scanner.py

```diff
@@ -124,27 +87,65 @@ def _publish_result(
         )
         self._publish(message)
 
+    def _start_scan(
+        self,
+        host_ip: str,
+        host_name: str,
+        installed_packages: Iterable[Package],
+        package_advisories: PackageAdvisories,
+    ) -> Generator[PackageVulnerability, None, None]:
+
+        for package in installed_packages:
+            package_advisory_list = (
+                package_advisories.get_package_advisories_for_package(package)
+            )
+            for package_advisory in package_advisory_list:
+                if package_advisory.package > package:
+                    yield PackageVulnerability(
+                        host_ip=host_ip,
+                        host_name=host_name,
+                        package=package,
+                        fixed_package=package_advisory.package,
+                        advisory=package_advisory.advisory,
+                    )
+
     def run_scan(
         self,
         message: ScanStartMessage,
     ) -> None:
         """Handle the data necessary to start a scan,
         received via mqtt and run the scan."""
 
-        installed_packages = [
-            RPMPackage.from_full_name(name) for name in message.package_list
-        ]
-        scan = NotusScan(self._loader)
+        package_advisories = self._loader.load_package_advisories(
+            message.os_release
+        )
+        if not package_advisories:
+            logger.info("No advisories found for %s", message.os_release)
+            return None
+        package_type = package_advisories.package_type
+
+        may_installed: Iterable[Optional[Package]] = (
+            DEBPackage.from_full_name(name)
+            if package_type == PackageType.DEB
+            else RPMPackage.from_full_name(name)
+            for name in message.package_list
+        )
```

Comment:
Just one small improvement to avoid executing the if clause in every loop. IMHO the Iterable[Optional[Package]] typing should also be obsolete. Sadly the suggestion is not formatted correctly.
```suggestion
        package_class = DEBPackage if package_type == PackageType.DEB else RPMPackage
        may_installed: Iterable[Optional[Package]] = [package_class.from_full_name(name) for name in message.package_list]
```

Reply:
yes the typing was for me since pyright went crazy before filtering